sync echo effect to BPM

Be-ing opened the pull request:
make Delay knob range from 0 to 2 beats

Be-ing committed:
sync echo effect to BPM

make Delay knob range from 0 to 2 beats

ronso0 commented:
Cool, thanks for working on this and #1254 and #1255!!
So the knob would be linear, meaning center position would be 1 beat?

Be-ing commented:
>  So the knob would be linear, meaning center position would be 1 beat?

Kinda but not exactly. The minimum is not 0, so the exact center of the knob is approximately 1 beat but not exactly 1 beat. I'll try implementing a workaround for that.

ronso0 commented:
Okay. But the default (right-clicking the knob) can be set to 1 beat, right?

Be-ing commented:
> Okay. But the default (right-clicking the knob) can be set to 1 beat, right?

I have the default set as 1/2 beat. Try it, I think you'll find it makes sense.

daschuer commented:
We cannot just replace delay by beats. Echo is a typical mic effect without any beat information.

Be-ing commented:
> We cannot just replace delay by beats. Echo is a typical mic effect without any beat information.

As explained by the comment in the code, a tempo of 100 BPM is assumed if there is no tempo information available. Do you have any ideas for how to handle that case better?

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into tempo_sync_echo

Be-ing committed:
rename Echo parameter Delay to Time, use 0-2 sec range without BPM

and make parameter knob center exactly 1 beat / 1 second

Be-ing commented on src/effects/native/echoeffect.cpp:
```diff
@@ -26,14 +26,16 @@ EffectManifest EchoEffect::getManifest() {
+    delay->setName(QObject::tr("Time"));
+    delay->setDescription(QObject::tr("Delay time\n"
+        "0 - 2 beats if tempo is detected (decks and samplers) \n"
+        "0 - 2 seconds if no tempo is detected (mic & aux inputs, master mix, headphone mix)."));
```
@esbrandt is this a good way to handle the tooltip?

ronso0 commented:
I tested this together with #1254 (which works great with Echo!) but somehow the default delay is always a bit higher than twice the BPM.
To test I played 2-beat loops with clear high hats or bass drums, Time knob I left at default position (1/2 beat?). For a 160 BPM track the echo had around 180 BPM, and for a 100 BPM track I got someting like 108.

ronso0 commented:
And Sent was full On after start although Meta knob was a lower limit.

Be-ing commented:
> I tested this together with #1254 (which works great with Echo!) but somehow the default delay is always a bit higher than twice the BPM.
To test I played 2-beat loops with clear high hats or bass drums, Time knob I left at default position (1/2 beat?). For a 160 BPM track the echo had around 180 BPM, and for a 100 BPM track I got someting like 108.

Maybe you're getting confused by the feedback? Try playing with the feedback parameter. And yes, the default time is 1/2 beat.

ronso0 commented:
> Maybe you're getting confused by the feedback? Try playing with the feedback parameter.

May be, but I don't see how feedback affects timing. This sounded odd that's why I measured it.
Will test again.

Be-ing committed:
add sync button to echo effect

In case someone does not want it to synchronize to the tempo

Be-ing committed:
round synced echo time to 1/4 beats

Be-ing commented:
Do you like the new scaling for the Time parameter? It ranges from 1/4 to 2 beats and the value is floored to the nearest 1/4 beat. If we like this, I'll adopt it for the other temporal effects after merging this. I think flooring to the nearest 1/2 beat might be more appropriate for flanger and phaser.

daschuer commented on src/effects/native/echoeffect.cpp:
```diff
@@ -26,26 +27,28 @@ EffectManifest EchoEffect::getManifest() {
-    delay->setDescription(QObject::tr("Delay time (seconds)"));
+    delay->setName(QObject::tr("Time"));
+    delay->setDescription(QObject::tr("Delay time\n"
+        "1/4 - 2 beats rounded down to nearest 1/4 beat if sync parameter is enabled and tempo is detected (decks and samplers) \n"
```
1/4 - 2 beats, snapping to the nearest 1/4 beat step if ...

daschuer replied:
It is a pity that the current value is only part of the debug tooltip. In this case it would be handy to have this in normal mode as well.

Be-ing replied:
It is visible without running in developer mode.

daschuer replied:
I mean the "current value of the CO like "1.5". This is not visible.

Be-ing replied:
Oh, yeah that is a shame. But let's leave that issue for after 2.1.

daschuer commented on src/effects/native/echoeffect.cpp:
```diff
@@ -26,26 +27,28 @@ EffectManifest EchoEffect::getManifest() {
 
     EffectManifestParameter* delay = manifest.addParameter();
     delay->setId("delay_time");
-    delay->setName(QObject::tr("Delay"));
```
Delay is IMHO the correct name for this parameter and should be kept. Time is only the physical quantity, without further meaning.

Be-ing replied:
I think "Delay" is confusing because that often refers to a more complex effect:

> The term “echo” was used more often in the early days, and is sometimes used today to refer to the distinct and distant repeats of a signal, while “delay” refers to anything from the same, to the short repeats heard as reverb, to the complex, long, manipulated repeats of an intricate digital delay line.

http://www.gibson.com/News-Lifestyle/Features/en-us/effects-explained-echo-delay.aspx

daschuer replied:
OK.In this case "Time" is OK.

daschuer commented on src/effects/native/echoeffect.cpp:
```diff
@@ -26,26 +27,28 @@ EffectManifest EchoEffect::getManifest() {
-    delay->setName(QObject::tr("Delay"));
-    delay->setDescription(QObject::tr("Delay time (seconds)"));
+    delay->setName(QObject::tr("Time"));
+    delay->setDescription(QObject::tr("Delay time\n"
```
"Delay time", is redundant, because a delay is always a time.

daschuer commented:
I have just tested this on top of master. 
It looks like the echo still saves its buffer over a enabled/disabled cycles. Haven't we fixed this earlier?

daschuer commented on src/effects/native/echoeffect.cpp:
```diff
@@ -26,26 +27,28 @@ EffectManifest EchoEffect::getManifest() {
-    feedback->setMaximum(1.0);
+    feedback->setMinimum(0.25);
+    feedback->setDefault(0.75);
+    feedback->setMaximum(0.90);
```
This should stay on 1.0. This allows to cycle a echo until infinity, after we have no input anymore.

daschuer replied:
I think there is also a use-case for feedback 0.0. this allows to fade the effect out or in from zero.

Be-ing replied:
That is what the send knob is for.

Be-ing replied:
> This should stay on 1.0. This allows to cycle a echo until infinity, after we have no input anymore.

This is an echo effect, not a looper. The effect should not be able to continue forever. I tried setting the maximum feedback to 0.99 and it prevents the output level from continuously escalating and getting out of hand until the deck clips, but the output continues for a very long time and effectively acts as a loop.

daschuer replied:
Do we have any complains about it? I think not, so this is not a required change for this PR and for me a undesired regression. 

Clipping can happen anyway with electronic music and an exact beat loop. 
We should consider to replace the clipper with a tanh clipper to make it sound more natural (more analog). 
https://github.com/mixxxdj/mixxx/blob/master/src/effects/native/phasereffect.cpp#L147
or better the fast approximation: 
https://github.com/mixxxdj/mixxx/blob/75e85d359d964ce9b55c92ebe3e7ed5962b51947/src/engine/enginefiltermoogladder4.h#L238

Be-ing replied:
> Do we have any complains about it?

I am complaining about it. The effect's buffer clipping isn't the issue; the ease of clipping the deck with feedback at 1, send at 1, and dry/wet at 1 is the issue.

daschuer replied:
That seams to be a known issue of an echo/delay effect, and sometimes a desired feature. 

> Each one of these repetitions will gradually fade out as new ones are created, unless you have the feedback cranked up full, in which case the signal will keep compounding until it reaches a deafening squall.

http://www.musicradar.com/tuition/tech/the-ultimate-guide-to-effects-delay-457920

I have just read some stompbox manuals and there is no other feedback range than 0 .. 100 % implemented. 

I think we have an other issue here with our gain leveling. If we play a track with RG loudness, the digital level is likely at 50 %, our delay buffer clamper on the other hand, clamps on digital full level, which is most likely much louder than the source signal and will clip the soundcard output with any tiny other signal added. 

If we use a tanh clipper this would be better because it kicks in somhow earlier. We can also add a gain to the clipping value to let et clip even earlier and may also make this a parameter which is default to a value that matches the input level.

Be-ing replied:
> That seems to be a known issue of an echo/delay effect, and sometimes a desired feature.

Hmm, if it's the standard way I could revert it. Anyone else have an opinion?

> I think we have an other issue here with our gain leveling. If we play a track with RG loudness, the digital level is likely at 50 %, our delay buffer clamper on the other hand, clamps on digital full level, which is most likely much louder than the source signal and will clip the soundcard output with any tiny other signal added.
If we use a tanh clipper this would be better because it kicks in somhow earlier. We can also add a gain to the clipping value to let et clip even earlier and may also make this a parameter which is default to a value that matches the input level.

That's an interesting idea. The output of the effect is divided by 2 before mixing with the dry signal. Let's not delay merging this PR for that though. If you have ideas for improving that, let's merge this first, then you can open a new PR for that.

Be-ing replied:
I did a bit more research, and you are right, a restricted range for the feedback parameter is not normal on an echo effect. I have restored it to the 0 - 1 range with 0.75 as default. I came up with an interesting alternate metaknob linking that somewhat approximates a beatmasher effect: link both the Time and Feedback parameters to the metaknob, invert the link for the Time parameter, unlink the Send parameter and turn it all the way up.

daschuer commented:
It look like there is something wrong with the Echo effect.
Test: play just one beat from a track, using a feedback of one and a delay of 2 beats. 
Expected behavior: The beat should loop at halve tracks bpm. 
Actual behavior: I hear two beats with different tempos cycling arround. 

Does the beat go back to 2 s after the track has stopped? I think this should stick to the current track bpm as it were playing. 
In an earlier version, it is possible to change the BPM of the single beat by the delay knob. This is now somehow broken as well.

daschuer commented on src/util/math.h:
```diff
@@ -65,6 +65,13 @@ inline int roundUpToPowerOf2(int v) {
     return power;
 }
 
+inline double roundToFraction(double value, int denominator) {
```
we should round to the nearest value, otherwise all values have a "snaping region", but not the topmost value,
2 in this case, is only adopted on the very maximum.

daschuer commented:
I think "Delay" is confusing because that often refers to a more complex
effect.

Ah, yes, I have just checked echo stomp boxes and they use "time" or "echo"
for this knob. Since echo also names the effect on some  these boxes,
"time" is just fine. Sorry for the noise.

Be-ing commented:
> Does the beat go back to 2 s after the track has stopped? I think this should stick to the current track bpm as it were playing.

No, it does not. I tested adding a `qDebug() << groupFeatures.beat_length` and watching the debugging output. The beat length is still reported to the effect correctly when the deck reaches the end of the track.

Be-ing commented:
> I have just tested this on top of master.

I suggest testing it with #1254.

> It looks like the echo still saves its buffer over a enabled/disabled cycles. Haven't we fixed this earlier?

I cannot reproduce this. I merged the fix in master into this branch.

Be-ing committed:
make roundToFraction round to nearest instead of clamp down

Be-ing committed:
echo effect: allow 1/8 beats as minimum delay time

Be-ing committed:
echo effect: don't snap to 1/4 seconds without beat syncing

1/4 seconds are totally arbitrary and sound weird without beat
synchronization

Be-ing commented:
I've changed the scaling of the Time parameter a bit. The minimum value is now 1/8 beat or 1/8 second, which produces a "slapback" effect. Above this minimum, it snaps to the nearest 1/4 beat. The change from 1/8 to snapping to nearest 1/4 is slightly strange, but I think it is okay. I tried snapping to the nearest 1/8 beat, but above 1/8, multiples of 1/8 beats (that aren't reducible to 1/4 beats) feel weird. I removed the snapping to nearest 1/4 second without tempo sync because it was totally arbitrary and didn't really make sense.

Be-ing committed:
echo effect: restore feedback parameter range to 0 - 1

Be-ing commented:
@ferranpujolcamins, could you test this?

Be-ing commented:
Is there anything left to do here? Ready for merge?

Be-ing commented on src/effects/native/echoeffect.h:
```diff
@@ -26,14 +26,14 @@ struct EchoGroupState {
             : delay_buf(mixxx::AudioSignal::kSamplingRateMax * kMaxDelaySeconds *
                         kChannelCount) {
```
With kMaxDelaySeconds as 2, this will not be big enough when the Time parameter is at its maximum and the BPM is below 60. Should kMaxDelaySeconds be increased? Is that edge case worth the higher memory consumption? These buffers consume a lot of memory, especially with #1254.

Be-ing replied:
I'll increase it now that I've come up with a solution for the ridiculous memory usage in #1254.

Be-ing committed:
increase size of Echo effect buffers

I doubt anyone will try to use Echo with the Time parameter at
the maximum of 2 beats with a tempo lower than 40 BPM.

daschuer commented:
The beat sync is out of beat. 
* Load two the same track to both decks.
* Enable sync and Quantize on both decks
* Start one deck. 
* Enable echo on the second deck with time = 1 and feedback = 1 
* enable echo and play just one beat. 
This beat should now cycle with the original BPM . 
Unfortuanely it does not and is soon off from the other deck still palying along. 

Another test:
* Play a on beat loop
* enable echo with time = 1 and feedback = 1
The beat should get louder and louder. Instead you hear a machine gun sound after a while.

Be-ing commented:
Is it the beat syncing that is out of beat or the feedback? Is the effects system providing the correct value with the beat_length on the groupFeatures?

Be-ing committed:
Echo effect: fix artifact when feedback is 1

Be-ing commented:
> enable echo with time = 1 and feedback = 1
The beat should get louder and louder. Instead you hear a machine gun sound after a while.

I don't entirely understand why, but somehow the combination of setting feedback to 1 and the ramping applied to the signal being stored in the echo buffer caused this artifact.

daschuer commented on src/effects/native/echoeffect.cpp:
```diff
@@ -95,42 +110,63 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
-                    / EchoGroupState::kRampLength;
+        // When feedback is 1.0, continously applying ramping to the buffer creates
+        // an unpleasant artifact.
+        if (feedback_amount != 1.0) {
```
It looks like the ramping code is wrong here. From the theory, an echo is a steady stream, echoing the original stream with a delay. This means that we do not apply a ramp if the effect is settled. What we need is a ramp that is a applied to the first samples in the buffer. I think that can be done by ramping the send_amount from zero after starting the effect. Changes in send_amount and feedback_amount is required anyway, to avoid clicks when changing.

Be-ing replied:
That seems like a better idea. I was having trouble understanding why the end of the period was being ramped.

mevsme commented:
Hi. I have a few questions about this FX:
* why ping-pong 100% right repeat one "sample" two times every channel, shouldn't it repeat it one time? And I see that Freq is bind to ping-pong so answer is not "because of such a frequency". Is it hard coded into the effect?
* why send 100% right still sounds quieter than the original (dry) sound? In Cubase DAW (I think in others too) 100% mix means that DRY sound you will not be header at all (for example if the effect is applied to a 1-quoter audio sample in 4/4 bar), you will not hear dry sound, only wet sound when time to play delayed sound comes. 50% mix means sample and its echoes are of equal level at firs playback (and if feedback is 100%) than it sound infinity as loud as initial sample. But in mix is sound like 60% dry and 40% wet sound.

Be-ing committed:
Echo: ramp send parameter

Be-ing committed:
Echo: move state maintenance code to end of process function

IMO the code is easier to understand this way.

Be-ing commented:
Okay, I have corrected the ramping of the send parameter. This fixes the strange machine gun sounding artifact with feedback fully up. Is there anything left to do for this? Ready for merge?

Be-ing commented:
@mevsme:
> why ping-pong 100% right repeat one "sample" two times every channel, shouldn't it repeat it one time?

The ping pong keeps going as long as there is sound in the effect's buffer, which is dependent on the send and feedback and parameters.

> why send 100% right still sounds quieter than the original (dry) sound?

Do you have the chain dry/wet knob fully wet? For the full volume of the effect, both the effect's send parameter and the chain dry/wet knob need to be fully up.

mevsme commented:
@Be-ing
Yes, I understand PingPong mechanics, maybe I'm not clear enough. I have max. feedback, portion of sound in buffer and deck stopped. So PingPong sounds L L R R L L R R... not L R L R L R.

Does it depends on buffer's sample length?

Be-ing commented:
> I have max. feedback, portion of sound in buffer and deck stopped. So PingPong sounds L L R R L L R R... not L R L R L R.

I cannot reproduce this.

Be-ing commented:
Ping. Let's finish this up. Are there any remaining issues?

mevsme commented:
I am not online yet. Tomorrow I will make a small video what I am talking about.

Be-ing commented:
@daschuer ready for merge?

@mevsme if there is indeed an issue with the ping-ponging, this PR has not modified that code at all so it is not related.

mevsme commented:
@Be-ing no, I don't think it's an issue, just quibbles

Be-ing commented:
Is there anything left to do here? Ready for merge?

daschuer commented:
The syncing back-end is broken. It is simply not in sync in any case.
My idea was to fix it along with: 
https://github.com/mixxxdj/mixxx/pull/1335
and than adopt the changes here and merge it. 

I also not sure if the snapping values and falling back to second when not synced is the right approach.
Testing with the similar Autopan effect changes, it turns out that the current implementation is not usable for me. 

How about scale the knob by beats in any case, beats are there. This will allow to unsync the effect without a change a significant tempo change. In that turn we can also reduce the snap positions to lets say five per turn e.g.: 1/2 1 2 3 4 or even more to just 1 / 3 / 4. 
This allows to hit the desire value quite better. All exotic fractions can be set manually, which should be possible, because we stick with a bpm scale an not switch to seconds at all.

Be-ing commented:
> The syncing back-end is broken. It is simply not in sync in any case.

Can you elaborate? As far as I can tell it works fine when the engine sample rate matches the file's sample rate (which I think is most cases, but some sound cards only support 48 kHz). I have been using it for a while now and it works well for me.

> I also not sure if the snapping values and falling back to second when not synced is the right approach.

I'm not sure if falling back to seconds when the SYNC parameter is off or beat info isn't available is the best approach either, but I don't know what would be better.

> How about scale the knob by beats in any case, beats are there. This will allow to unsync the effect without a change a significant tempo change. In that turn we can also reduce the snap positions to lets say five per turn e.g.: 1/2 1 2 3 4 or even more to just 1 / 3 / 4.
This allows to hit the desire value quite better. All exotic fractions can be set manually, which should be possible, because we stick with a bpm scale an not switch to seconds at all.

I'm not sure I understand what you're proposing. Are you proposing to change the behavior so that the time parameter is rounded to the nearest fraction when SYNC is on but when SYNC is off and there is tempo information available, the time parameter scale is continuous but uses the same range of beats?

daschuer commented:
Yes.

daschuer commented:
Is https://github.com/mixxxdj/mixxx/pull/1256#issuecomment-313547489 fixed in the meanwhile? I have not tested it again lately.

Be-ing commented:
> Yes

Which question was that in response to?

> Is #1256 (comment) fixed in the meanwhile? I have not tested it again lately.

Yes I fixed that in 9bcd2de.

daschuer commented:
Ah, cool, I will re-test. 

My single yes was for scaling the knob by beats even if sync is disabled.

Be-ing commented:
> My single yes was for scaling the knob by beats even if sync is disabled.

Okay, I will try that approach. If we end up adopting that, the "SYNC" parameter should be renamed.

Be-ing commented:
> How about scale the knob by beats in any case, beats are there.

Would would a use case of this behavior be? What use is a 0.6734 beat echo versus a 0.50 beat echo? As the SYNC parameter is currently implemented, it can be turned off to manually set an echo time in case the detected tempo is wrong. I do not think your proposed behavior would help that use case. It may actually be counterproductive for that use case.

Be-ing committed:
Echo: add Triplet parameter

This can be used to add some intentional wonkiness.

Be-ing committed:
Echo: remove SYNC parameter

I do not think there is a use case for an unsynced Echo effect.

Be-ing commented:
I have removed the SYNC parameter because it is impractical to set the Time manually while performing, hence the need for this PR in the first place. If the tempo is not detected correctly, the effect is not very useful IMO.

I have added a parameter to switch between rounding to 1/4 or 1/3 beats. In a future PR I'd like to add high & low pass filters to the effect. If the Sync parameter was kept and those were added, the parameters would get really cluttered.

Be-ing committed:
Echo: update Time parameter description

Be-ing commented:
This PR has been open over 4 months. I'd really like to finish this and move on already. Is there anything left to do here?

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into tempo_sync_echo

Be-ing committed:
Echo: fix Time depending on sample rate

Be-ing committed:
Echo: fix scaling of Time parameter

daschuer commented:
Ok, an other test an syncing works now great.

What is the use case for the triplet switch?
Triplets are 3 notes played in the time of two. The most famous triplets track is the start wars theme 
|  da   da | ta ta ta |  da  da |
I think that is not what you are aiming for.
It looks more like you targeting a 3/4 track, right? Shouldn’t the knob label be "3/4" ? 

Why allow only 4/4 and 3/4 tracks? There are so many other measures. 

The best would be to store the measure along with the beat grid and adopt it by the effect. 
Can't we move the measure info out of the effect region, into the deck? This is required by so many other Mixxx facilities, so it is probably to short to target this in a single effect. 

What do you think?


There is an implementation issue, when I go from a short echo to long. Instead of sound, you hear the gaps for the time offset. 
In along times, an echo effect was created by a two head tape deck, were you could shift the play head away from the record head. Such a solution does not suffer from the silence issue after increasing the time. I think we should cycle over the whole buffer and not do "% delay_samples" to fix this.  


Since you like to have this merged soon, I think we should go back to the sync (snapping) knob to allow to have an echo for all measures. the gap issue can be fixed in a separate PR.

Be-ing commented:
The Triplet parameter is unrelated to the beats per measure. The use case is to make intentionally choppy sounding echoes.

I noticed the issue with increasing the Time parameter leaving silence in the buffer as well. This is not a new issue and unrelated to the changes in this PR, so I agree it should be saved for another PR. IMO it is a minor issue because I don't think there is a use case for increasing the Time parameter while the effect is audible.

Be-ing commented:
> Triplets are 3 notes played in the time of two.

That is not correct. Triplets are 3 notes in the time of one beat. How many beats there are in the measure is separate.

Be-ing commented:
> Triplets are 3 notes in the time of one beat.

Researching more, apparently we're both correct. A triplet can be three notes in the time of any power of 2 note (1/8, 1/4, 1/2, 1, 2, 3...)

daschuer commented:
Here is a good explanation: https://en.wikipedia.org/wiki/Tuplet

What is the use case for your current triplets knob?

Be-ing commented:
The best use case for triplet echoes I have identified so far is intentionally creating a disorienting tension between the echoes and the dry signal with the volume fader up. For mixing between tracks I do not think triplets work very well.

daschuer commented:
OK, I have played with it a bit. 

The major question is: can we effort to always have the snapping behavior? 

I think not, because we loose the smooth changing of the delay as effect.
A mandatory feature of a Tape Delay effect. On the other hand the internal as a Tape Delay is broken, because of the silence gaps. 
   
Triplets: 
You are right, triplets are unrelated to the measure, because they are defined by notes and not by beats. 
If we look at the by far most common triplet type, the 1/8T (three notes played instead of two eighth notes)
you are also right that it fits exactly it one beat. 

If we look for example to a 1/4T, it can change the beat signature of a track. A 4/4 measure played with triplets can be recognized as a 6/4 measure. That is also the reason can divide our beat-grid by 2/3. 

Expressed in fractions of an 1/4 beat, we currently have these snapping points: 
Triplets off:  
0.125  1/32 note
0.25    1/16 note
0.5      1/8 note  
0.75    1/8 dotted
1         1/4 note
1.25
1.5      1/4 dotted 
1.75
2         1/2 note 

Triplets on:  
0.166  1/16 triplet
0.33    1/8 triplet
0.66    1/4 triplet 
1         1/4 note
1.33    1/2 triplet
1.66    
2         1/2 note 

for me, these points are hard to guess on one hand and if we consider other triplets than 1/8T, we have also to consider dotted noted on the other hand. 

like here: 
https://github.com/mixxxdj/mixxx/pull/1301
https://www.samplemagic.com/journal/2017/03/chord-design-techniques/

.. this can be overdone. 

If we snap all points that can be expressed by single notes, we have 15 Snap points, which are quite hard to handle. I think the Maximum we can guess by looking at the knob position is 5. So maybe we should pick 5 reasonable of them. 

The other issue is that a triplet or dotted rhythm is an anomaly of a normal rhythm. from this point of view it could be handy to switch with them instantly without touching the knob. 
So we can consider to always snap to a power of two fraction of a beat and multiply them by 2/3 afterwards, if you enable the triplet knob. So you can change form a beat echo (1/4 note) to a 1/4T echo instantly.

What do you think?

Be-ing commented:
> The major question is: can we effort to always have the snapping behavior? I think not, because we loose the smooth changing of the delay as effect.

You are right. I played with my friend's analog delay pedal and it is quite fun to play with wiggling the time knob. Then the remaining question is when the snapping switch is off and tempo information is available, should the time knob still scale to beats or to a fixed scale of seconds? I think the effect implementation has to be fixed to try each approach to make a decision on that question.

> change form a beat echo (1/4 note) to a 1/4T echo instantly.

That's an interesting idea. I'll play with it.

Be-ing commented:
Good point. Suddenly switching the time value when toggling the sync/snap parameter would be strange.

Be-ing committed:
Echo: add Quantize parameter, change behavior of Triplet parameter

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into tempo_sync_echo

Be-ing commented:
I added back a parameter to toggle between rounding to beats and continuous scaling. Because the Time is always synced to the beat, I am calling the parameter Quantize rather than Sync. It does not automatically snap to beat boundaries (yet. Would someone like to take that task up later?)

I changed the behavior of the Triplet parameter. Instead of rounding to 1/3 beats, it rounds to 1/4 beats and multiples that by 1/3. So, when the knob is at center and the Triplet parameter is off, it is 1 beat (1/4 note in a 4/4 time signature track). When the Triplet parameter is turned on, it switches to 1/3 beat. This way, the position of the switch points on the knob's range are the same regardless of whether the Triplet parameter is on. I tried multiplying by 2/3 instead of 1/3, but IMO it feels more intuitive this way with 1/3.

Be-ing commented:
Anyone else care to test? @ronso0?

ronso0 commented:
> Anyone else care to test? @ronso0?

Sure, if I find time to compile this week

Be-ing commented:
Ardour ships with a set of LV2 plugins including a delay which has the desired effect when adjusting its time parameter. For reference, the code of the effect is here: https://github.com/Ardour/ardour/blob/0b6e2d1e46b9dfe72fe879d2ddddafeafda07023/libs/plugins/a-delay.lv2/a-delay.c#L448

Be-ing commented:
I copied the algorithm from Ardour's a-Delay plugin. It seems to work, but I don't fully understand all the details. There may still be room for improvement.

Be-ing commented on src/effects/native/echoeffect.cpp:
```diff
@@ -149,90 +147,108 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
+            }
+
+            gs.wet_sample_left = SampleUtil::clampSample(
+                    (gs.wet_sample_left + gs.delay_buffer[read_position_left]) / 2.0);
```
a-Delay does some curious [crossfading between the samples here](https://github.com/Ardour/ardour/blob/master/libs/plugins/a-delay.lv2/a-delay.c#L459) using `1 / sampleRate`, but I don't understand the reasoning behind that.

daschuer replied:
Without fully understanding the code, that might be for the pitch effect, of e real tape delay if you change the delay time. Without it we might here click sounds.

Be-ing replied:
There are some minor click sounds when changing the Time parameter quickly but I am not sure of their source. I tried the crossfading with 1/numSamples technique that a-Delay uses but it does not solve the issue. The size of the engine's buffer is an artifact of the software's configuration and shouldn't have any impact on the sound of an effect.

Be-ing replied:
Emulating the pitch shifting of an analog delay would be fun, but I don't know how to implement that.

Be-ing replied:
Guitarix has a nice sounding tape delay LV2 plugin but it is written in Faust and does some complicated modeling of tape including wow & flutter. I checked a-Delay again and adjusting its Time parameter does not have the popping sound this does. I'm not sure what is different that is causing the popping here.

daschuer commented on src/effects/native/echoeffect.cpp:
```diff
@@ -95,88 +122,133 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
+        // period is a number of beats
+        if (m_pTripletParameter->toBool()) {
+           if (period <= 0.125) {
+              period = 0.041666666666666664; // 1/8 / 3
```
Use her directly "1/8 / 3".  The preprocessor is able to calculate that at compile time.

daschuer commented:
The parameter interface is nice now. 
Did you consider to reduce the quantization points on the delay knob?

Be-ing commented:
Do you mean to round to 1/8 beats or to 1/2 beats?

daschuer commented:
The range is OK. I originally thought to halve the delay on every step, like we do with the looping controls. 
The real issue is, that we have no visual feedback if we run in non developer mode.

Be-ing commented:
Yes, that is definitely an issue. But solving the visual feedback issue would require quite some more work to make a new skin widget, update the skins, and create a way for the effect to show a nicely formatted value (like 1/4 instead of 0.25223134). IMO that can wait for the following release.

Be-ing commented:
Although we do not have visual feedback of the current value of parameters, we can at least specify the ranges in the descriptions so the user can have a clue what is going on by reading the tooltip.

Be-ing commented:
If you would prefer, I can split the last commit off to a new branch so we can merge this now.

daschuer commented:
Yes good idea.The interface is nice now, so the buffering changes are more easy to review in a separate PR. 
Can you fix the looong decimals issues in sis PR?

Be-ing committed:
Echo: code cleanup

Be-ing committed:
Echo: correct parameter descriptions

Be-ing commented:
Ready for merge?

daschuer commented:
Thank you, this is a real fun effect now.
LGTM.

Be-ing commented:
Great! I will test the syncing with flanger and phaser again. I am not sure if the Quantize and Triplet parameters make sense for those effects like they do for Echo.

ronso0 commented:
Is it intended that echo fades out much faster with Triplets enabled?
Would be handy to have the same fade time like without it.

Be-ing commented:
Yeah, the echo time is divided by 3. If you want to extend the fade out, you could turn up the feedback.

daschuer commented:
Adopting the normal fade out time for triplets is IMHO a valid request. Unfortunately the feedback gain behaviour might be unexpected in case it is adopted for triplets. 
Mmm, what is the most expected behaviour?
